### PR TITLE
Replace all get_device() functions with get_device_from_id() or get_device_from_array()

### LINKED ISSUE #2545: Do not use deprecated APIs internally
We need not use deprecated APIs internally. Remove them and check if we use them.

### PR DESCRIPTION
**Pull request by mitmul**
Fix #2545 

Note that the default value of `device_id` argument in `chainer/cuda.py` at L:156 is None and L:163 in the same file has been changed to `if type(device_id) in _integer_types` because this function could take `None` to get `DammyDevice` object and it should return `DammyDevice` also for NumPy's scalar type like `numpy.int64(0)`.

### CODE REVIEW

File: chainer/cuda.py [hunk 1]
```diff
@@ -153,14 +153,14 @@ def __ne__(self, other):
 # ------------------------------------------------------------------------------
 # Global states
 # ------------------------------------------------------------------------------
-def get_device_from_id(device_id):
+def get_device_from_id(device_id=None):
```
Comment (unnonouno) [inline chainer/cuda.py]: Is this default value required?

File: chainer/cuda.py [hunk 2]
```diff
@@ -153,14 +153,14 @@ def __ne__(self, other):
...
     """Gets the device from an ID integer.
 
     Args:
         device_id (int or None): The ID of the device which this function
             returns.
     """
-    if device_id is not None:
+    if type(device_id) in _integer_types:
```
Comment (unnonouno) [inline chainer/cuda.py]: Why is this change required?

File: docs/source/tutorial/gpu.rst [hunk 3]
```diff
@@ -106,29 +106,33 @@ It is equivalent to the following code using CuPy:
...
    :func:`chainer.cuda.to_cpu` and :func:`chainer.cuda.to_gpu` functions automatically switch the current device correctly.
 
-Chainer also provides a convenient function :func:`chainer.cuda.get_device` to select a device.
-It accepts an integer, CuPy array, NumPy array, or None (indicating the current device), and returns an appropriate device object.
-If the argument is a NumPy array, then *a dummy device object* is returned.
-The dummy device object supports *with* statements like above which does nothing.
-Here are some examples:
+Chainer also provides a convenient function :func:`chainer.cuda.get_device_from_id` and :func:`chainer.cuda.get_device_from_array` to select a device.
+The former function accepts an integer or None.
+When None is given, it returns *a dummy device object*.
```
Comment (unnonouno) [inline docs/source/tutorial/gpu.rst]: <pre>``None``</pre>

Comment (mitmul) [inline docs/source/tutorial/gpu.rst]: OK, I'll fix it and also fix L:110

File: chainer/cuda.py (hunk 2 above)
Comment (mitmul) [inline chainer/cuda.py]: See my first post

**Comment by mitmul** [comment]
As for the necessity of `if device else None`, for example, `to_gpu` could take `None` as the input argment named `device`, so in that case, `device` does not have `id` property then the condition is needed.

**Comment by unnonouno** [comment]
Is that really expected behavior?
```it should return DammyDevice also NumPy's scalar type like numpy.int64(0).```
In this case, user should call `get_device_from_array`?

**Comment by mitmul** [comment]
https://github.com/pfnet/chainer/pull/2546/files#diff-f77865af49e320e789c79d6300ca9702L38

This test `test_get_device_for_numpy_int` seems to intend to check that it surely returns `DammyDevice` when the given integer is a numpy's scalar type of numpy.int64. So, I just tried to make it to follow the same behavior. Was the original test before this PR wrong? What is the desirable behavior? It seems to be depending on how do we think what is 'integer'

**Comment by unnonouno** [comment]
We need to consider when a user call `get_device` with numpy types. When shape of ndarray is `()`, operators for this value sometimes returns numpy types. 
```
>>> x = numpy.array(1, 'f')
>>> type(x)
<type 'numpy.ndarray'>
>>> type(x + x)
<type 'numpy.float32'>
```
In suche case, a user calls `get_device_for_array` because if shape of an input array is not `()`, its result is also an ndarray.  So, what we need to do is to fix `get_device_for_array` isn't it?

File: chainer/cuda.py (hunk 2 above)
Comment (mitmul) [inline chainer/cuda.py]: Thank you for the internal discussion! I'll fix it :)

**Comment by unnonouno** [comment]
Check Travis result.

**Comment by unnonouno** [comment]
Check parameter.py L27. This fix finds that parameter link has a bug.

**Comment by mitmul** [comment]
Thank you for pointing out it. I didn't notice it. In the line, `cuda.get_device_from_array(array)` should be given to `to_gpu` method. I fixed it :+1:

**Comment by unnonouno** [comment]
Check test result.

File: chainer/link.py [hunk 4]
```diff
@@ -328,7 +328,9 @@ def to_gpu(self, device=None):
         if not self._cpu:
             return self
         d = self.__dict__
-        with cuda.get_device_from_id(device.id if device else None):
+        if not device_id:
```
Comment (unnonouno) [inline chainer/link.py]: `not 0` is `True`. Use `device_id is None`.

Comment (unnonouno) [inline chainer/link.py]: And can you make a test for such a case?  I mean a test which calls `to_gpu(0)` with calling `use` with device `1`.

File: chainer/variable.py [hunk 5]
```diff
@@ -215,15 +215,15 @@ def to_cpu(self):
...
 
         Args:
-            device: Target device specifier. If omitted, the current device is
-                used.
+            device_id (int): The target device ID. If omitted, the current
+                device is used.
 
         """
-        with cuda.get_device_from_id(device.id if device else None):
+        with cuda.get_device_from_id(device_id):
```
Comment (unnonouno) [inline chainer/variable.py]: Ah sorry. This behavior is changed. Is it really required?

File: chainer/link.py (hunk 4 above)
Comment (mitmul) [inline chainer/link.py]: OK, I added two tests.

File: chainer/variable.py (hunk 5 above)
Comment (mitmul) [inline chainer/variable.py]: I found `get_device` function are still remaining in `to_gpu()` method defined in `chainer/cuda.py`, so I'll replace them with `get_device_from_id` and `get_device_from_array`. Then the below two `to_gpu` will also be replaced with `to_gpu(array, device_id)`.

**Comment by mitmul** [comment]
@unnonouno Fixed. Thank you for reviewing.

File: tests/chainer_tests/test_link.py [hunk 6]
```diff
@@ -98,6 +98,18 @@ def test_to_gpu(self):
         self.assertIsInstance(self.link.y.grad, cupy.ndarray)
         self.assertIsInstance(self.link.p, cupy.ndarray)
 
+    @attr.gpu
+    def test_to_gpu_different_device(self):
+        cuda.Device(1).use()
```
Comment (unnonouno) [inline tests/chainer_tests/test_link.py]: Change the current device after the test using `setUp` and `tearDown`.

File: chainer/link.py [hunk 7]
```diff
@@ -310,16 +310,16 @@ def to_cpu(self):
         self._device_id = None
         return self
 
-    def to_gpu(self, device=None):
+    def to_gpu(self, device_id=None):
```
Comment (unnonouno) [inline chainer/link.py]: In the original implementation, `to_gpu` method accept `Device` object as an argument. We can't change this behavior.

Comment (mitmul) [inline chainer/link.py]: Ah, sorry I confused with v2. I'll fix it, thanks!

**Comment by mitmul** [comment]
All Jenkins tests passed

File: tests/chainer_tests/test_link.py (hunk 6 above)
Comment (mitmul) [inline tests/chainer_tests/test_link.py]: I'm afraid I'm not sure what you mean, but I added tearDown method to reset the current device.

Comment (unnonouno) [inline tests/chainer_tests/test_link.py]: Make it work without GPUs.

**Comment by mitmul** [comment]
Fixed setup to use cuda.cupy... only when cuda.available is True

**Comment by unnonouno** [comment]
LGTM. Please merge the master

**Comment by unnonouno** [comment]
OK, LGTM
